src/rebar_app_discover.erl

```diff
@@ -305,8 +300,19 @@ find_app(AppInfo, AppDir, Validate) ->
 %% the directories where source files can be located. Returns the related
 %% app info record.
 -spec find_app(rebar_app_info:t(), file:filename_all(),
-               [file:filename_all()], valid | invalid | all) ->
+               [file:filename_all()], valid | invalid | all, rebar_state:t()) ->
     {true, rebar_app_info:t()} | false.
+find_app(AppInfo, AppDir, SrcDirs, Validate, State) ->
+    AppInfo1 = case ec_file:real_dir_path(rebar_dir:root_dir(State)) of
+                   AppDir ->
+                       Opts = rebar_state:opts(State),
+                       rebar_app_info:default(rebar_app_info:opts(AppInfo, Opts), Opts);
```

Comment:
alright I guess this is the killer bit -- if the app's path is the same as the project root, use the pre-loaded state, otherwise consult. Am I reading this correctly?

Reply:
Right. And note that this function is only called for project apps. So it isn't like it has to do this for dependencies.